### PR TITLE
Normalizer / Denormalizer organization

### PR DESCRIPTION
**Pull request by nidup**
baby PR to discuss about normalizers / denormalizers

### CODE REVIEW

File: UPGRADE-1.5.md [hunk 1]
```diff
@@ -190,6 +190,59 @@ akeneo_storage_utils:
...
+    └── Structured          -> format "json", "xml", used to generate internal standard format (product template values, product draft values), or for rest api, can also be used with the updater api
+
+Pim/Bundle/EnrichBundle
+└── Normalizer              -> format "internal_api", used by the internal rest api to communicate with new UI Forms (product edit form)
+
+We could have [WIP],
+
+Pim/Bundle/TransformBundle/Normalizer/Structured -> Pim/Component/Catalog/Normalizer/Structured
+
+Pim/Bundle/TransformBundle/Normalizer/Flat -> Pim/Component/Connector/Normalizer/Flat
```
Comment (juliensnz) [inline UPGRADE-1.5.md]: IMHO as we have ArrayToStandardConverter, those normalizer/denormalizer should not even exist (we just need to implement StandardToArrayConverter for export)

File: UPGRADE-1.5.md [hunk 2]
```diff
@@ -190,6 +190,59 @@ akeneo_storage_utils:
...
+ -> because should only be used for csv import/export (versioning for legacy reasons)
+
+Pim/Bundle/CatalogBundle/MongoDB/Normalizer -> Pim/Bundle/CatalogBundle/Normalizer/MongoDB/Document
```
Comment (juliensnz) [inline UPGRADE-1.5.md]: why not in a component ?

File: UPGRADE-1.5.md [hunk 3]
```diff
@@ -190,6 +190,59 @@ akeneo_storage_utils:
...
+Pim/Component/Catalog/Normalizer/MongoDB/NormalizedData -> Pim/Bundle/CatalogBundle/Normalizer/MongoDB/NormalizedData
+ -> to normalize MongoDB in CatalogBundle because glue to Mongo
```
Comment (juliensnz) [inline UPGRADE-1.5.md]: ok, I understand

Comment (juliensnz) [inline UPGRADE-1.5.md]: Maybe normalizer is not a good usage here. We should do Object -> standard format -> Mongodb format. First step could be done by a standard normalizer and the second one could be done by a StandardToMongoConverter

File: UPGRADE-1.5.md (hunk 1 above)
Comment (nidup) [inline UPGRADE-1.5.md]: Denormalizer ~= arrayconverter + updater 
Normalizer are always use for object to array

Comment (juliensnz) [inline UPGRADE-1.5.md]: ah, yeah, you are right

Comment (pierallard) [inline UPGRADE-1.5.md]: You have the new Localizer Normalizers too!

File: UPGRADE-1.5.md [hunk 4]
```diff
@@ -190,6 +190,59 @@ akeneo_storage_utils:
             override: Acme\Bundle\AppBundle\Model\ProductValue
 ```
 
+## Normalizers & Denormalizers
+
+The PIM heavily uses the Serializer component of Symfony http://symfony.com/doc/2.7/components/serializer.html.
+
+Especially, classes which implement,
+ - NormalizerInterface to transform object to array
```
Comment (phaseinducer) [inline UPGRADE-1.5.md]: "transform objects into arrays" or "tranform an object into array"

File: UPGRADE-1.5.md [hunk 5]
```diff
@@ -190,6 +190,59 @@ akeneo_storage_utils:
...
+
+Especially, classes which implement,
+ - NormalizerInterface to transform object to array  
+ - DenormalizerInterface to transform array to object
+
+We have a lot of different formats and for backward compatibility reason, we never re-organized them.
+
+Pim/Bundle/CatalogBundle
+└── MongoDB
+    └── Normalizer          -> format "mongodb_json", used to generate the field normalized_data in a product mongo document
```
Comment (phaseinducer) [inline UPGRADE-1.5.md]: I think "normalized_data" is in camel case.

File: UPGRADE-1.5.md (hunk 2 above)
Comment (phaseinducer) [inline UPGRADE-1.5.md]: It's just used for normalizedData in Mongo, not sure it's relevant to put that in a component.

Comment (phaseinducer) [inline UPGRADE-1.5.md]: @nidup: the target namespace shouldn't be Pim/Bundle/CatalogBundle/Normalizer/MongoDB/NormalizedData here ?

File: UPGRADE-1.5.md (hunk 3 above)
Comment (phaseinducer) [inline UPGRADE-1.5.md]: I guess the line here is :
Pim/Bundle/TransformBundle/Normalizer/MongoDB -> Pim/Bundle/CatalogBundle/Normalizer/MongoDB/Document

File: UPGRADE-1.5.md (hunk 4 above)
Comment (nidup) [inline UPGRADE-1.5.md]: one object to one array :) 
but a doctrine Collection is an object, does not work with [] ;)
http://api.symfony.com/2.7/Symfony/Component/Serializer/Normalizer/NormalizerInterface.html

**Pull request merged**